### PR TITLE
fix: rebuild cycle history from verified period dates

### PR DESCRIPTION
**Pull request by TaiKamilla**
## 📌 Summary

Some users open the app and find their cycles tab showing only a single cycle (or none) even though their calendar still has all the period days they logged. The verified dates are intact in storage, but the prediction `history` array got truncated by an earlier sync, migration, or upgrade. This change rebuilds the missing history from the verified dates on login so the cycles tab matches what the user actually logged.

## 🔗 Ticket / Issue

* GitHub Issue: [#237](https://github.com/Oky-period-tracker/periodtracker/issues/237)

## ✅ Changes

* [x] Fix: `PredictionProvider` now reconciles `state.prediction.history` against `state.answer[userId].verifiedDates` once per logged-in user. When more period clusters are detected in verified dates than the engine has cycles, the prediction state is rebuilt and dispatched, then the engine takes over normally.
* [x] Fix: New utility `app/src/prediction/historyReconciliation.ts` exposes `detectPeriodClusters`, `shouldReconcileHistory`, `buildPredictionFromVerifiedDates`, and the higher-level `reconcileHistoryWithVerifiedDates`. Clusters allow up to a 1-day intra-period gap so a single forgotten log does not split a real period.
* [x] Tests: 8 unit tests covering cluster detection, the trigger heuristic, sparse-history rebuild, and engine-flag preservation. Verified period payloads are generated by a `makePeriods` helper instead of hardcoded fixtures.

## 🧪 Testing

Steps to reproduce the bug (manual):
1. Take a user account with rich verified dates but sparse `prediction.history` (e.g. seed the `oky_user.store` JSON with the payload from issue #237).
2. Log in as that user, open the cycles tab.
3. Before the fix: only the cycles known to `state.prediction.history` render.
4. After the fix: every cycle implied by the verified dates renders, with the most recent period as the current cycle.

Automated:
* `cd app && yarn jest --testPathPattern=historyReconciliation` runs the new suite.
* The `rebuilds the missing cycles when history is sparse` test seeds 12 generated periods, leaves history empty, and asserts 11 history entries plus the current cycle.

* ✅ Expected result: cycles tab lists every cycle the user logged.
* ❌ Bugs to watch out for: very fragmented period logs with consistent 3+ day gaps would be grouped into a single cycle. The current heuristic only merges runs separated by at most one missed day.

### Before

https://github.com/user-attachments/assets/1ee4e78d-c231-4310-955e-5586b139d1c6

### After

https://github.com/user-attachments/assets/859b2a22-d259-4096-8f2d-31c7bf6614a8

## 📖 Notes for Reviewers

* Reconciliation runs once per user via a ref keyed on user id, so logging out and logging back in as a different user re-triggers the check.
* The rebuild only fires when `verifiedDates` contains more period clusters than `history.length + 1`, so accounts whose history is already correct are unaffected.
* History entries are emitted newest-first to match the existing engine convention. Smart prediction circular buffers and SMA values are repopulated from the recovered cycle and period lengths.

---

### Checklist

* [x] Code follows style guidelines
* [x] Self-review completed
* [x] Tests added/updated
* [ ] Documentation updated
* [x] No sensitive info (keys, secrets, etc.) committed

### CODE REVIEW

File: app/src/contexts/PredictionProvider.tsx [hunk 1]
```diff
@@ -56,6 +64,22 @@ export function PredictionProvider({ children }) {
...
+    if (!currentUserId) return
+    if (reconciledUserRef.current === currentUserId) return
+    if (!predictionState?.currentCycle) return
+    if (!verifiedDates || Object.keys(verifiedDates).length === 0) return
+
+    const rebuilt = reconcileHistoryWithVerifiedDates(predictionState, verifiedDates)
+    if (rebuilt) {
+      reduxDispatch(actions.setPredictionEngineState(rebuilt))
+    }
+    reconciledUserRef.current = currentUserId
```
Comment (DantonMariano) [inline app/src/contexts/PredictionProvider.tsx:80]: One small suggestion: set reconciledUserRef.current = currentUserId before the dispatch to avoid a redundant effect re-fire.

**Review by DantonMariano — APPROVED** [review]
LGTM!
